<!DOCTYPE html>
<head>
<title>Population Genomics Viewer</title>
<link rel="shortcut icon" href="img/icon.png">
<link rel="apple-touch-icon" href="img/icon.png">
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

<link rel="stylesheet" href="incl/chosen/chosen.css">
<style type="text/css">
body {
    background-image: url(img/dna_background.png);
    background-repeat: repeat-x;
    background-position: 0 0px;
    padding-top: 30px;
}
</style>
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
<script type='text/javascript' src='incl/chosen/chosen.jquery.min.js'></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Viewer stuff -->
<script type='text/javascript' src='js/util.js'></script>
<script type='text/javascript' src='js/genome_track.js'></script>
<script type='text/javascript' src='js/data_row.js'></script>
<script type='text/javascript' src='js/horizontal_panel.js'></script>
<script type='text/javascript' src='js/vertical_panel.js'></script>
<script type='text/javascript' src='js/absolute_panel.js'></script>
<script type='text/javascript' src='js/text_draw_panel.js'></script>
<script type='text/javascript' src='js/tree_panel.js'></script>
<script type='text/javascript' src='js/button.js'></script>
<script type='text/javascript' src='js/remote.js'></script>
<script type='text/javascript' src='js/tracks.js'></script>
<script type='text/javascript' src='js/viewer.js'></script> <!-- this has to be last because it requires everything else -->

<script type='text/javascript'>
// state of the updating display
var show_updating;
var show_downloading;
var update_angle;

// compute location of ctrl point for arc drawing
var pt0 = (1,0);
var pt1 = (Math.cos(Math.PI*0.8), Math.sin(Math.PI*0.8));
var between = ((pt0[0]+pt1[0]) / 2, (pt0[1]+pt1[1]) / 2);
var ctrl_pt_dist = Math.sqrt(Math.pow(between[0], 2) + Math.pow(between[1], 2)) * 2;

// global start and end pos to display
var display_pos;

var viewer;
var metadata;

// url of the filtered file to download
var file;

function update() {
    var params = getParams();
    if(!params)
        return;
    viewer.update(params[0], params[1], params[2], params[3]);
}

// amt is a multiplier 
function zoom(amt) {
    viewer.zoomRatio(amt);
}

// amt is represented as a fraction of the visible field and is positive or negative indicating direction 
function pan(amt) {
    viewer.pan(amt);
}

/* -------------------------------------------------------
   Params
   -------------------------------------------------------
*/

function parsePos(num) {
    num = num.replace(',', '');
    var multiplier = 1;
    if(num.indexOf('M') != -1)
        multiplier *= 1000000 * (num.split('M').length - 1);
    if(num.indexOf('m') != -1)
        multiplier *= 1000000 * (num.split('m').length - 1);
    if(num.indexOf('K') != -1)
        multiplier *= 1000 * (num.split('K').length - 1);
    if(num.indexOf('k') != -1)
        multiplier *= 1000 * (num.split('k').length - 1);
    num = num.replace('k', '').replace('m', '').replace('K', '').replace('M', '');
    num = parseFloat(num);
    num *= multiplier;
    return Math.floor(num);
}

function getStrains() {
    var selects = $('.chzn-select');
    var strains = [];
    for(var a in selects) {
        var sel = selects[a];
        if(sel.childNodes) {
            for(var i = 0; i < sel.childNodes.length; i++) {
                if(sel.childNodes[i].selected || selected_sets[a]) { // if selected, or whole set selected
                    strains.push(sel.childNodes[i].value); 
                }
            }
        }
    }
    if(strains.length == 0) { // no strains selected
        $('.chzn-select').parent().parent().addClass('error');
    }
    else {
        $('.chzn-select').parent().parent().removeClass('error');
    }
    
    if(strains.length > 0)
        return strains;
    return null;
}

function getStartEnd() {
    var start = parsePos(document.getElementById('start_pos').value);
    var end = parsePos(document.getElementById('end_pos').value);
    
    if(isNaN(start)) {
        $('#start_pos').addClass('error');
        $('#start_pos').parent().parent().addClass('error');
    }
    else {
        $('#start_pos').removeClass('error');
        $('#start_pos').parent().parent().removeClass('error');
    }
    if(isNaN(end)) {
        $('#end_pos').addClass('error');
        $('#end_pos').parent().parent().addClass('error');
    }
    else {
        $('#end_pos').removeClass('error');
        $('#end_pos').parent().parent().removeClass('error');
    }
    if(!isNaN(start) && !isNaN(end)) {
        return [start, end];
    }
    return null;
}

function getChrom() {
    var chrom = document.getElementById('chrom').value;
    return chrom;
}

function getParams() {
    var range = getStartEnd();
    var strains = getStrains();
    var chrom = getChrom();
    if (!(range && strains && chrom)) {
        $('#error_box').show();
        return null;
    }
    else {
        $('#error_box').hide();
        return [chrom, range[0], range[1], strains];
    }
}


/* -------------------------------------------------------
   Updating
   -------------------------------------------------------
*/

function updating(width, height, x, y) {
    // the updating circle is 150px diameter
    update_angle = (update_angle + Math.PI/20) % (Math.PI*2);
    
    var canvas = document.getElementById('canvas');
    var c = canvas.getContext('2d');
    c.canvas.width  = width;
    c.canvas.height = height;
    var r = 75; // radius
    // clear only the circle
    c.clearRect(x,y,x+r*2,y+r*2);
    
    c.strokeStyle = "#000000";
    c.fillStyle = "#000000";
    c.lineWidth = 15;
    
    c.beginPath();
    c.arc(x+r, y+(r+12), r, update_angle, update_angle + Math.PI*0.8, false)
    c.stroke();
    c.closePath();
    
    c.beginPath();
    c.moveTo(x+r + Math.cos(update_angle + Math.PI*0.8) * (r+12), y+(r+12) + Math.sin(update_angle + Math.PI*0.8) * (r+12));
    c.lineTo(x+r + Math.cos(update_angle + Math.PI*0.8) * (r-12), y+(r+12) + Math.sin(update_angle + Math.PI*0.8) * (r-12));
    c.lineTo(x+r + Math.cos(update_angle + Math.PI*0.8+0.2) * r, y+(r+12) + Math.sin(update_angle + Math.PI*0.8+0.2) * r);
    c.fill();
    c.closePath()
    
    c.beginPath();
    c.arc(x+r, y+(r+12), r, update_angle + Math.PI, update_angle + Math.PI*1.8, false)
    c.stroke();
    c.closePath();
    
    c.beginPath();
    c.moveTo(x+r + Math.cos(update_angle + Math.PI*1.8) * (r+12), y+(r+12) + Math.sin(update_angle + Math.PI*1.8) * (r+12));
    c.lineTo(x+r + Math.cos(update_angle + Math.PI*1.8) * (r-12), y+(r+12) + Math.sin(update_angle + Math.PI*1.8) * (r-12));
    c.lineTo(x+r + Math.cos(update_angle + Math.PI*1.8+0.2) * r, y+(r+12) + Math.sin(update_angle + Math.PI*1.8+0.2) * r);
    c.fill();
    c.closePath()

    c.font = '50px sans-serif';
    c.textBaseline = 'top';
    c.fillText('Updating...', x-20, y+r*2+12);
}

/* -------------------------------------------------------
   Whole genome
   -------------------------------------------------------
*/

function toggle_full_set(event) {
    var $selector = event.data.selector;
    var $checkbox = event.data.checkbox;
    var i = event.data.i;
    if(!selected_sets[i]) {
        $checkbox.removeClass('glyphicon-unchecked');
        $checkbox.addClass('glyphicon-check');
        
        // disable the chzn selector
        $selector.attr('data-placeholder', 'ALL SELECTED');
        $selector.attr('disabled', false);
        $selector.trigger("chosen:updated");
        
        selected_sets[i] = true;
    }
    else {
        $checkbox.removeClass('glyphicon-check');
        $checkbox.addClass('glyphicon-unchecked');
        
        // enable the chzn selector
        $selector.attr('data-placeholder', 'Choose samples');
        $selector.attr('disabled', false);
        $selector.trigger("chosen:updated");
        
        selected_sets[i] = false;
    }
}

// from twitter.github.com/bootstrap/appliation.js (bootstrap doc specific code)
// fix sub nav on scroll
!function ($) {
  $(function(){
    var $win = $(window)
    var $nav = $('.subnav')
    var navTop = $('.subnav').length && $('.subnav').offset().top - 40
    var isFixed = 0

    processScroll()

    $win.on('scroll', processScroll)

    function processScroll() {
      var i, scrollTop = $win.scrollTop()
      if (scrollTop >= navTop && !isFixed) {
        isFixed = 1
        $nav.addClass('subnav-fixed')
      }
      else if (scrollTop <= navTop && isFixed) {
        isFixed = 0
        $nav.removeClass('subnav-fixed')
      }
    }
  })
}(window.jQuery)

/* -------------------------------------------------------
   setup
   -------------------------------------------------------
*/
function setup() {
	file = "meta." + shareholder + ".json";
	$.getJSON(file, function(data) {
        metadata = data;
        viewer = new Viewer(document, 1, metadata);
        document.getElementById('tracks').appendChild(viewer.element);
        addDescriptions(metadata);
        addSets(viewer.sets);
        addChroms(viewer.chroms);
        $(".chzn-select").chosen();
    });
}

function addDescriptions(metadata) {
  $("#overview").append(metadata.description);
  for(var t in metadata.tracks) {
    $("#datasets").append("<b>" + metadata.tracks[t].name + "</b><br/>" + metadata.tracks[t].description + "<br/><br/>");
    $("#datasets").append("Color mapping: ");
    for(var c in metadata.tracks[t].colors) {
      var color = metadata.tracks[t].colors[c];
      $("#datasets").append("<span style='color:rgba(" + color[0] + "," + color[1] + "," + color[2] + "," + color[3] + "); margin-right:10px;'>" + c + "</span>");
    }
    $("#datasets").append("<br/><br/>");
  }
}

function addChroms(chroms) {
    $select = $('#chrom');
    for(var chrom in chroms) {
        $option = $('<option>' + chroms[chrom].display + '</option>');
        $option.attr('value', chrom);
        $select.append($option);
    }
    dispBounds(1, chroms[$('#chrom').val()].length);
}

function dispBounds(st, en) {
  $('#start_pos').val(st);
  $('#end_pos').val(en);
}

var selected_sets = [];

function addSets(sets) {
    $div = $('#strain_sets');
    var i = 0; // set index
    for(var set_name in sets) {
        selected_sets.push(false);
        $div.append($('<br>'));
        
        $control = $('<div>');
        $control.addClass('input-group');
        // select
        $select = $('<select>');
        $select.addClass('chzn-select');
        $select.attr('multiple', 'true');
        $select.attr('name', 'strains');
        $select.attr('id', set_name + '_strains');
        $select.attr('data-placeholder', 'Choose samples');
        $select.css('width', '500px');
        for(var j = 0; j < sets[set_name].length; j++) {
            $option = $('<option>' + sets[set_name][j] + '</option>');
            $option.attr('value', sets[set_name][j]);
            $select.append($option);
        }
        
        $label = $('<label>');
        $label.addClass('input-group-addon');
        $label.append(set_name + ' ');
        // pick all button
        $all = $('<i>');
        $all.addClass('glyphicon glyphicon-unchecked');
        $all.css('cursor', 'pointer');
        $all.click({selector: $select, checkbox: $all, i: i}, toggle_full_set);
        $label.append($all);
        
        $control.append($label);
        $control.append($select);
        $div.append($control);
        i++; // set index
    }
}

</script>
</head>
  
	
	<?php
	$user = $_SERVER['REMOTE_USER'];
        //declare LDAP 
	$ldaphost = '';
	$ds = ldap_connect($ldaphost) or die( "failed to connect : LDAP {$ldaphost}" );
	ldap_set_option($ds, LDAP_OPT_PROTOCOL_VERSION, 3);
        //declare  Distinguished Name and ldap password
        $ldapbind = ldap_bind($ds, '', '');
        //declare  Distinguished Name
	$dn = "";
        //build request to extract user's hierarchical lvl 
	$filter = "";
	$member = array("cn");
	$sr=ldap_search($ds, $dn, $filter, $member);
	$info = ldap_get_entries($ds, $sr);
        //get group 
	$groupLDAP = $info[0]['cn'][0];
        if (!in_array($groupLDAP.".json", scandir("./services"))){
	   // no dedicated config file, switch to demo mode
           $groupLDAP="demo";
        }
	?> 
	<script type="text/javascript">
		var shareholder ='<?PHP echo $groupLDAP;?>';
	</script>
	<body onload='setup()'>
    <div class="container">

      <header class='page-header'>
        <div style="display:table; width:100%">
        <div style="display:table-row">
            <div style='display:table-cell; text-align:left'>
                <h1>Population Genomics Viewer V2.0</h1>
            </div>
        </div> <!-- /table-row -->
        </div> <!-- /table -->
        <p class='lead'>
          Comparative genome browser
        </p>
        <p>
         For comments or questions, send mail to <a href="mailto:">mailto@support.com</a>
        </p>
        <div class='subnav' id='subnav'>
          <ul class='nav nav-tabs'>
            <li><a href="#description">Description</a></li>
            <li><a href="#viewer">Viewer</a></li>
            <li><a href="#reference">Reference</a></li>
          </ul>
        </div>
      </header>
      
      <div class="well" style='margin-top:40px'>
      <section id='description'>
				<p>
					<h4>PGV presentation</h4>
					The Population Genomics Viewer (PGV) is a custom genome browser designed to provide the user answers to questions on the haplotype diversity and phylogenetic origin of the genetic variation underlying any genomic region (adapted from MPV: Wang et al. 2012). 
					<br/><br/>
					<h4>Selection of a genomic region</h4>
					PGV allows the user to select a region of the genome and a set of samples (lines). 
					<br/>
					The genomic region is selected by specifying the chromosome (e.g. "Chr1"), the start (e.g. "31200000" or "31200K" or "31.2M"), and end of the interval.
					Samples can be selected by name or by entire set.
					<br/>
					The selection of the whole dataset takes precedence over the selection of the lines
					<br/>
					If the selection of data failed, (while moving Forward/Backward, ZoomIN ZoomOUT ...) then click again on the "View filter data" button.
					<br/><br/>
					<h4>Visualization and dynamic functions</h4>
					To allow users to further customize the visualization to their needs, PGV  support dynamic recoloring of haplotypes blocks (Wang et al. 2012; Figure 4) as well as dynamic sorting of samples at a user selected position (Wang et al. 2012; Figure 5). Dynamic coloring and reordering tools facilitate comparison of features by visually aligning regions where genomes are similar and different. To zoom in, you just need to click and drag across a track to highlight a region, and then click Zoom.
					<br/><br/>
					<h4>Download button</h4>
					Each track present a download button which allows the user to download the shown data.
					<br/><br/>
          <h4>Experiment Overview</h4>
          <div id="overview">
          </div>
          <br/>
          <h4>Data display settings</h2>
          <div id="datasets">
          </div>
        </p>
      </section>
      </div>

      <div class="well">
      <section id='viewer'>
        <h3>Viewer</h3>
        <p>
          <br/>
          Samples and subset selection
          <br/>
          <br/>
          <form class='form-horizontal' action='dump.zip' method='POST' id='download_form'>
            <fieldset>

              <legend>Filter</legend>
              
              <div class="input-group">
                <label class='input-group-addon' for="chrom">Chromosome</label>
                  <select class="form-control" name="chrom" id="chrom">
                  </select>
              </div>
              
              <div class="input-group">
                <label class='input-group-addon' for="start_pos">Start</label>
                <input class="form-control" id="start_pos" name="start_pos" type="text"/>
              </div>
              <div class="input-group">
                <label class="input-group-addon" for="end_pos">End</label>
                <input class="form-control" id="end_pos" name="end_pos" type="text"/>
              </div>
              <div class="input-group" id='strain_sets'>
              </div>

              <div class="input-group">
                <div class="controls">
                  <p class="help-block">
                    Type a sample name to search. Multiple samples can be selected by Ctrl+Click.
                    <br/>
                    Click the box next to the set name to select all.
                  </p>
                  
                  <p class="alert alert-error" id='error_box' style='width:400px; display:none;border-color: red; text-align: center;'>
                    <font color ="red"> <strong>/!\ WARNING /!\</strong> <br/>Please select panel or specific line </font>
                  </p>
                  <br/>
                  <a onclick='update()' class='btn btn-primary' id='view_filter_button'><i class="glyphicon glyphicon-refresh"></i> View filter data</a>
                </div>
              </div>
              
              <br/>
              
              <div class="control-group">
                <label class="control-label" for="end_pos">Navigation</label>
                <div class='controls'>
                  <a onclick='zoom(0.1)' class='btn btn-default'><i class='glyphicon glyphicon-zoom-out'></i> 10x</a>
                  <a onclick='zoom(0.5)' class='btn btn-default'><i class='glyphicon glyphicon-zoom-out'></i> 2x</a>
                  <a onclick='zoom(2)' class='btn btn-default'><i class='glyphicon glyphicon-zoom-in'></i> 2x</a>
                  <a onclick='zoom(10)' class='btn btn-default'><i class='glyphicon glyphicon-zoom-in'></i> 10x</a>
                  <br/>
                  <br/>
                  <a onclick='pan(-0.5)' class='btn btn-default'><i class='glyphicon glyphicon-chevron-left'></i><i class='glyphicon glyphicon-chevron-left'></i></a>
                  <a onclick='pan(-0.2)' class='btn btn-default'><i class='glyphicon glyphicon-chevron-left'></i></a>
                  <a onclick='pan(0.2)' class='btn btn-default'><i class='glyphicon glyphicon-chevron-right'></i></a>
                  <a onclick='pan(0.5)' class='btn btn-default'><i class='glyphicon glyphicon-chevron-right'></i><i class='glyphicon glyphicon-chevron-right'></i></a>
                </div>
                <div class='controls'>
                  <p class="help-block">
                    To zoom in, click and drag across a track to highlight a region, then click [Zoom]
                  </p>
                </div>
              </div>
              
            </fieldset>
          </form>
          
          <legend>Tracks</legend>
          <div id='tracks'>
          </div>
          
        </p>
      </section>
      </div>
      <div class="well" style='margin-top:40px'>
        <section id="reference">
          <h2>Reference</h2>
          <p>
            Adapted from:
            <br>
            <a href ="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3311102/pdf/1471-2105-13-S3-S13.pdf">Jeremy R. Wang, Fernando Pardo-Manuel de Villena, and Leonard McMillan. Comparative analysis and visualization of multiple collinear genomes. BMC Bioinformatics, 2012. </a>
          </p>
        </section>
      </div>
      <hr>
      <footer class="footer">
        <p>
          <img src="img/unc_logo.gif">
          &nbsp;&nbsp;&nbsp;&nbsp;
          <img src="img/jax_logo.png">
          &nbsp;&nbsp;&nbsp;&nbsp;
          <img src="img/CGD_logo.png" style="width:200px">
          &nbsp;&nbsp;&nbsp;&nbsp;
        </p>
      </footer>
      <hr>
    </div><!-- /container -->

</body>
</html>
